### ISSUE TITLE
Concat arrays inside deepProperties and deepStaticProperties

### ISSUE THREAD

**Issue by koresar** [stampit-org/stamp-specification#69]
We should reconsider using the "lodash/merge" function. Let's just use `mergeWith`: https://lodash.com/docs#mergeWith

Let me copy&paste the example from the link above:
```js
function customizer(objValue, srcValue) {
  if (_.isArray(objValue)) {
    return objValue.concat(srcValue);
  }
}

var object = {
  'fruits': ['apple'],
  'vegetables': ['beet']
};

var other = {
  'fruits': ['banana'],
  'vegetables': ['carrot']
};

_.mergeWith(object, other, customizer);
// → { 'fruits': ['apple', 'banana'], 'vegetables': ['beet', 'carrot'] }
```

So, my proposal: do not add any new Descriptor Properties. But replace the "merge" with the "mergeWith" code above.

* Not inflating the specs.
* Not complicating implementation.
* I have never used arrays inside the ".merge.*" object, it used to be useless.
* I'm not talking about "properties" or "staticProperties".
* It's about deeply merged items - "deepProperties" and "deepStaticProperties".
* Deep merging of two arrays currently happens index-by-index:
```js
let deep1 = compose({deepProperties: { key: ["a", "b", "c"] } });
let deep2 = compose({deepProperties: { key: [42, 42] } });
let composed = compose(deep1, deep2);
console.log(composed.compose.deepProperties); // { key: [ 42, 42, 'c' ] }
```
In other words, array indexes are like POJO keys.
```js
let deep1 = compose({deepProperties: { key: {0: "a", 1: "b", 2: "c"} } });
let deep2 = compose({deepProperties: { key: {0: 42, 1: 42} } });
let composed = compose(deep1, deep2);
console.log(composed.compose.deepProperties); // { key: { '0': 42, '1': 42, '2': 'c' } }
```
I find it useless. For the past 9 months since stampit v2 was released I have never ever used arrays inside the deepProperties just because of that.
* However, I was constantly finding that *array concatenation* would help me from time to time.

**Proposal:** arrays inside the `deepProperties` and `deepStaticProperties` should be concatenated instead of merged.

**Comment by unstoppablecarl** [comment]
I think this is a good idea but needs some implementation adjustments.

With the code above:
```js
let obj1 = { foo: ['a','b','c'] };
let obj2 = { foo: {bar: 'bam'} };

_.mergeWith(obj1, obj2, customizer) // { foo: [ 'a', 'b', 'c', { bar: 'bam' } ] } oops!
```

Proposed implementation

```js
function merge(obj1, obj2) {
    return _.mergeWith(obj1, obj2, function(v1, v2) {

        const v1Arr = _.isArray(v1);
        const v1Obj = _.isObject(v1);

        const v2Arr = _.isArray(v2);
        const v2Obj = _.isObject(v2);

        if (v1Arr && v2Arr) {
            return [].concat(v1, v2);
        } 
        else if ((v1Obj && v2Arr) || (v1Arr && v2Obj)) {
            return v2;
        }
    });
}

// wrote some tests

let obj1 = { foo: { bar: 'bam' } }
let obj2 = { foo: [ 'a', 'b', 'c' ] }
merge(obj1, obj2) // { foo: [ 'a', 'b', 'c' ] }

let obj1 = { foo: [ 'a', 'b', 'c' ] }
let obj2 = { foo: { bar: 'bam' } }
merge(obj1, obj2) // { foo: { bar: 'bam' } }

let obj1 = { foo: [ 1, 2, 3 ] }
let obj2 = { foo: [ 'a', 'b', 'c' ] }
merge(obj1, obj2) // { foo: [ 1, 2, 3, 'a', 'b', 'c' ] }

**Comment by koresar** [comment]
I gave your code a deep thought, tested in back and forth.

Your implementation have a bug - `return v2;`. This will return reference, which must never happen with `deepProperties`.

Here is my proposal of merging:
```js
const _ = require('lodash');

function merge(obj1, obj2) {
    const res = _.mergeWith(obj1, obj2, (v1, v2) => {
        if (Array.isArray(v1) && v2 !== undefined) return v1.concat(v2);
        if (Array.isArray(v2) && v1 !== undefined) return [v1].concat(v2);
    });
    console.log(res);
}

merge({foo: {bar: 'bam'}}, {foo: ['a']});   // { foo: [ { bar: 'bam' }, 'a' ] }
merge({foo: ['a']}, {foo: {bar: 'bam'}});   // { foo: [ 'a', { bar: 'bam' } ] }
merge({foo: [1]}, {foo: ['a']});            // { foo: [ 1, 'a' ] }
merge({foo: [1]}, {foo: 'a'});              // { foo: [ 1, 'a' ] }
merge({foo: 'a'}, {foo: [1]});              // { foo: [ 'a', 1 ] }
merge({foo: undefined}, {foo: [1]});        // { foo: [ 1 ] }
merge({foo: [1]}, {foo: undefined});        // { foo: [ 1 ] }
```

##### Why this is a little better.
Stamps need exactly this. The following simple code becomes possible:
```js
const AggregatingConveters = compose({ 
  staticDeepProperties: { 
    converter: [] // having this empty array will allow to collect here everything!
  }
});

const MyStampWithConverter = compose({ 
  staticDeepProperties: {
    converter: function () { /* implementation */ }
  }
});

let s1 = compose(AggregatingConveters, MyStampWithConverter); // THESE TWO LINES
let s2 = compose(MyStampWithConverter, AggregatingConveters); // YIELD SAME STAMP.

console.log(s1.compose.staticDeepProperties.converter.length); // 1
console.log(s2.compose.staticDeepProperties.converter.length); // 1
```
I find that behavior consistent. The most consistent and handy for stamps as possible.

**Comment by unstoppablecarl** [comment]
You raise some good points. 

I agree with the returning a right hand reference problem. I should have thought of that.

I do not agree with your proposed implementation.

I think the proposed behavior complex, unexpected, and hard to explain. 

Currently when there are 2 values that you cannot reasonably compose like a string and an int, the right value replaces the left. To me composing arrays and objects together should behave the same way as they are similarly incompatible.

```js
merge({foo: 'a'}, {foo: 1}); // { foo: 1 }
```

It is reasonable to assume that a number should replace the string value and things will work fine as they can be coerced into the same type. So whatever is reading that value could reasonably still function. For the most part the same is true with boolean and null values.

In most cases I think composing an array into a non-array would be done mistakenly. It may not be, and to be consistent I think the behavior should be to replace with the right hand value. This will also make it easier to locate the problem if the replacement was made in error.

I think the only clear case to assume the intent to concat 2 values is when they are both arrays.

I do not think composition symmetry important. The spec currently cannot maintain symmetry in most cases. 
```js
let a = { foo: 1 };
let b = { foo: 'a' };
_.merge(a, b); // { foo: 'a' }
_.merge(b, a); // { foo: 1 }
```

**Comment by unstoppablecarl** [comment]
Proposed implementation:

```js

function merge(obj1, obj2) {
    return _.mergeWith(obj1, obj2, function(v1, v2) {

        const v1Arr = _.isArray(v1);
        const v1Obj = _.isObject(v1);

        const v2Arr = _.isArray(v2);
        const v2Obj = _.isObject(v2);

        if (v1Arr && v2Arr) {
            return [].concat(v1, v2);
        } else if (v1Obj && v2Arr) {
            return [].concat(v2);
        } else if (v1Arr && v2Obj) {
            return merge({}, v2);
        }
    });
}
```

I wrote up tests describing exactly how this should behave https://gist.github.com/unstoppablecarl/492a3e770a69955ac6eb

Do you think the values of the right hand array should be deep copied?

**Comment by koresar** [comment]
Ok. Tests are good. I'll make a PR out of your gist. Looks like you found the most consistent way so far.

**Comment by koresar** [comment]
Hey, @unstoppablecarl. I've shortened your `merge()` function code. It works exactly the same. Your tests pass just fine. See it [here.](https://github.com/stampit-org/stamp-specification/pull/75/files#diff-1b3ef822ec87668673adf6a07c81d02cR12)

**Comment by unstoppablecarl** [comment]
Looks great.

You didn't answer my question:
> Do you think the values of the right hand array should be deep copied?

**Comment by unstoppablecarl** [comment]
https://github.com/stampit-org/stamp-specification/pull/75/files#diff-9ed4d701899cd250bab0c301ddab7cb2R119

In the tests, this part is not useful without displaying the `a` and `b` values. You should either remove it or add the a, b values to the message. Probably best to just remove them.

I am unsure if implementing the `concat` in `configuration` is the best choice. I would use it that way but I do not know if others would. What do you think?

**Comment by koresar** [comment]
Sorry I missed you question. Was to excited. :)
> Do you think the values of the right hand array should be deep copied?

Good question! Let's talk about it. :) What do you reckon?

> I am unsure if implementing the concat in configuration is the best choice. I would use it that way but I do not know if others would. What's your opinion?

I'm pretty sure `configuration` will benefit form `concat` even more than the `deepProperties`.

**Comment by unstoppablecarl** [comment]
I'm undecided about deep copying. That is why I asked. 

`_.merge` currently does not copy anything from the right value by ref. With the implementation we just agreed upon it would copy the values of arrays by ref. This would be inconsistent and surprising I think. But not all objects can be easily deep cloned like DOM nodes etc see https://lodash.com/docs#clone . Although you should not be trying to deep merge those things (Is there a valid case you would?). Thoughts?

Proposed implementation:

 ```js
const merge = (dst, src) => mergeWith(dst, src, (dstValue, srcValue) => {
   if (Array.isArray(dstValue)) {
     if (Array.isArray(srcValue)) return dstValue.concat(srcValue.map(_.cloneDeep);
     if (isObject(srcValue)) return merge({}, srcValue);
   }
 });
```

With `configuration`:

If you wanted to have an array that did not concat but be replaced in `configuration` what would your options be to accomplish that? I guess you could use `staticProperties` right? but then they are exposed statically.

This makes me wonder if we do need `staticInitializers` to give you the option to do whatever.
 
What do you think of applying the proposed implementation above and implementing `staticInitializers` as we discussed in other issues? I think that would allow the ability to do anything you need.

**Comment by koresar** [comment]
Unfortunately, this implementation will not allow us to collect functions:
```
if (Array.isArray(srcValue)) return dstValue.concat(srcValue.map(_.cloneDeep);
```
and here is why:
```
$ npm i n_ -g
$ n_
n_ > function F(){}
undefined
n_ > _.cloneDeep(F);
{}
```
So, we are not cloning array items.

**Comment by koresar** [comment]
> If you wanted to have an array that did not concat but be replaced in configuration what would your options be to accomplish that?

An object with numeric keys will suffice. `{0: value}`

**Comment by koresar** [comment]
> What do you think of ... implementing `staticInitializers` as we discussed in other issues?

`staticInitializers` are not needed anymore since we have this - [.compose replaceable via staticProperties](https://github.com/stampit-org/stamp-specification/issues/71)

**Comment by koresar** [comment]
Let's move this discussion to the #75. Closing this issue.

**Comment by koresar** [comment]
> If you wanted to have an array that did not concat but be replaced in configuration what would your options be to accomplish that?

I'm thinking of it again. There are multiple ways to sort that out.
* Use `properties` or `staticProperties` instead.
* Replace the ".compose" with `{staticProperties: newCompose}` and do it there.
* Clear it with `null` (`compose({configuration: {foo: null}})`) and reset the array with `[]` (`compose({configuration: {foo: []}})`). Then concat your array.

I'm thinking if we should introduce `deepConfiguration`... :)
